src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs

```diff
@@ -0,0 +1,282 @@
+                                !methodSymbol.Name.EndsWith(MandatoryAsyncSuffix, StringComparison.Ordinal) &&
+                                !HasAsyncCompatibleReturnType(methodSymbol, syncBlockingTypes))
+                            {
+                                string asyncMethodName = methodSymbol.Name + MandatoryAsyncSuffix;
+                                IEnumerable<IMethodSymbol> methodSymbols = semanticModel.LookupSymbols(
+                                    context.Operation.Syntax.GetLocation().SourceSpan.Start,
+                                    methodSymbol.ContainingType,
+                                    asyncMethodName,
+                                    includeReducedExtensionMethods: true)
+                                    .OfType<IMethodSymbol>();
+
+                                string containingMethodName = "";
+                                if (context.ContainingSymbol is IMethodSymbol parentMethod)
+                                {
+                                    containingMethodName = parentMethod.Name;
+                                }
+
+                                SyntaxNode invokedMethodName = context.Operation.Syntax;
+
+                                foreach (IMethodSymbol method in methodSymbols)
+                                {
+                                    if (!method.HasAttribute(systemObsoleteAttribute)
+                                        && HasSupersetOfParameterTypes(method, methodSymbol)
+                                        && method.Name != containingMethodName
+                                        && HasAsyncCompatibleReturnType(method, syncBlockingTypes))
+                                    {
+                                        Diagnostic diagnostic = invocationOperation.CreateDiagnostic(
+                                            Descriptor,
+                                            invokedMethodName.ToString(),
+                                            asyncMethodName);
+
+                                        context.ReportDiagnostic(diagnostic);
+
+                                        return;
+                                    }
+                                }
+                            }
+                        }
+                        else
+                        {
+                            InspectAndReportBlockingMemberAccess(context, syncBlockingSymbols, SymbolKind.Property);
+                        }
+                    }
+                }, OperationKind.Invocation, OperationKind.PropertyReference);
+            });
+        }
+
+        internal class SyncBlockingSymbol
+        {
+            public SyncBlockingSymbol(string Name, string Namespace, SymbolKind Kind, ISymbol Value)
+            {
+                this.Name = Name;
+                this.Namespace = Namespace;
+                this.Kind = Kind;
+                this.Value = Value;
+            }
```

Comment:
I'm not sure if a struct will be better or not, But was it considered?

Reply:
i am not sure either, so leaving as is